### PR TITLE
Optimize v3_kernel for thousands of clauses

### PR DESCRIPTION
**Pull request by josevalim**
Prior to this patch, v3_kernel would do multiple
passes on the clauses to group them. This commit
unrolls those passes, making v3_kernel up to 10%
faster in those cases.

### CODE REVIEW

File: lib/compiler/src/v3_kernel.erl [hunk 1]
```diff
@@ -2141,21 +2155,7 @@ arg_arg(Con) -> Con.
...
-	#k_nil{} -> k_nil;
-	#k_cons{} -> k_cons; 
-	#k_tuple{} -> k_tuple;
-	#k_map{} -> k_map;
-	#k_binary{} -> k_binary;
-	#k_bin_end{} -> k_bin_end;
-	#k_bin_seg{} -> k_bin_seg;
-	#k_var{} -> k_var
-    end.
+arg_con(Arg) -> element(1, arg_arg(Arg)).
```
Comment (kostis) [inline lib/compiler/src/v3_kernel.erl]: IMO this change is a code degradation.  The original code contains perfect type information to both the input argument and the result of the function while the new code is basically "anything goes".  Most notably, `#ialias{}` is now also handled (and probably in an unintended way) when in the original code it was not.

In general, I would advice _against_ such "optimizations"/code rewrites.   They are typically a very bad idea that come back to bite you at some point.

Comment (josevalim) [reply lib/compiler/src/v3_kernel.erl]: Yup. I was 80% sure that this change would be rejected. But I thought I would submit it to review anyway. :) I will revert it.

File: lib/compiler/src/v3_kernel.erl [hunk 2]
```diff
@@ -1590,19 +1590,12 @@ match_var([U|Us], Cs0, Def, St) ->
...
-match_con_1([U|_Us] = L, Cs, Def, St0) ->
+match_con([U|_Us] = L, Cs, Def, St0) ->
     %% Extract clauses for different constructors (types).
     %%ok = io:format("match_con ~p~n", [Cs]),
-    Ttcs0 = select_types([k_binary], Cs) ++ select_bin_con(Cs) ++
-        select_types([k_cons,k_tuple,k_map,k_atom,k_float,
-                      k_int,k_nil], Cs),
-    Ttcs = opt_single_valued(Ttcs0),
+    Ttcs0 = select_types(Cs, [], [], [], [], [], [], [], [], []),
+    Ttcs1 = [{T, Types} || {T, [_ | _] = Types} <- Ttcs0],
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/compiler/src/v3_kernel.erl]: It is not an assertion, we are filtering empty lists from Ttcs0. :)

File: lib/compiler/src/v3_kernel.erl [hunk 3]
```diff
@@ -1613,8 +1606,41 @@ match_con_1([U|_Us] = L, Cs, Def, St0) ->
...
+	    select_types(Cs, Bin, BinCon, Cons, Tuple, Map, [C | Atom], Float, Int, Nil);
+	k_float ->
+	    select_types(Cs, Bin, BinCon, Cons, Tuple, Map, Atom, [C | Float], Int, Nil);
+	k_int ->
+	    select_types(Cs, Bin, BinCon, Cons, Tuple, Map, Atom, Float, [C | Int], Nil);
+	k_nil ->
+	    select_types(Cs, Bin, BinCon, Cons, Tuple, Map, Atom, Float, Int, [C | Nil])
+    end;
+select_types([], Bin, BinCon, Cons, Tuple, Map, Atom, Float, Int, Nil) ->
+    [{k_binary, lists:reverse(Bin)}] ++ handle_bin_con(lists:reverse(BinCon)) ++
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/compiler/src/v3_kernel.erl]: Oh,  good catch. And `reverse` is even already imported. :+1:

**Comment by josevalim** [comment]
Feedback addressed.

File: lib/compiler/src/v3_kernel.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhogberg) [reply lib/compiler/src/v3_kernel.erl]: Ah right, I forgot how comprehensions work. :o)

**Comment by bjorng** [comment]
Thanks! Included in our daily builds.

**Pull request merged**